### PR TITLE
GraphNG: Stack by percentage

### LINKED ISSUE #36707: GraphNG: Add support for percent stacking
Grafan 8.0 introduced [possibility to stack series in the Time series panel](https://github.com/grafana/grafana/pull/30749). One big missing thing that we have not done is percent stacking.

We need to introduce percent stacking in order to have feature parity with the old Graph panel.

Relevant stacking code for Time series panel can be found here:  https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/TimeSeries/utils.ts#L252

### PR DESCRIPTION
**Pull request by leeoniya**
Fixes https://github.com/grafana/grafana/issues/36707

# 1500%!

still todo:

- [x] make sure we auto-set the axes/scale hard limits for stacked series to 0..1 or 0..100
- [x] make sure we auto-set units to %
- [x] figure out what's going on with the 1500% label display (looks like the displayProcessor is doing 15 x 100 when it detects % stacking mode, cause we feed it raw values and not ones in the 0...1 range)

how to handle stacking of series with mismatched units/scales?

negative stacking is still as messed up as before (no intent to address it in this PR).

image

### CODE REVIEW

File: packages/grafana-ui/src/components/uPlot/utils.ts [hunk 1]
```diff
@@ -62,18 +62,33 @@ export function preparePlotData(frame: DataFrame): AlignedData {
     seriesIndex++;
   }
 
+  const byPct = frame.fields[1].config.custom?.stacking?.mode === StackingMode.Percent;
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/uPlot/utils.ts:65]: i guess we have the stacking config editor disabled, so this is just fudge to use the same mode for all?

File: packages/grafana-ui/src/components/uPlot/config.ts [hunk 2]
```diff
@@ -285,6 +285,7 @@ export const graphFieldOptions = {
   stacking: [
     { label: 'Off', value: StackingMode.None },
     { label: 'Normal', value: StackingMode.Normal },
+    { label: 'Percent', value: StackingMode.Percent },
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/uPlot/config.ts:288]:
```suggestion
    { label: '100%', value: StackingMode.Percent },
```
to match what google calls it?

File: packages/grafana-ui/src/components/uPlot/utils.ts (hunk 1 above)
Comment (ryantxu) [reply packages/grafana-ui/src/components/uPlot/utils.ts:65]: interesting -- then maybe we should move the bar-chart stacking option back to field config :thinking: and avoid the extra copy/shift thing

File: packages/grafana-ui/src/components/uPlot/config.ts (hunk 2 above)
Comment (leeoniya) [reply packages/grafana-ui/src/components/uPlot/config.ts]: changed

File: public/app/plugins/panel/barchart/utils.ts [hunk 3]
```diff
@@ -268,6 +276,9 @@ export function prepareGraphableFrames(
             })
           ),
         };
+
+        copy.display = stacking === StackingMode.Percent ? getDisplayProcessor({ field: copy, theme }) : copy.display;
```
Comment (ryantxu) [inline public/app/plugins/panel/barchart/utils.ts:280]:
```suggestion
        if( stacking === StackingMode.Percent ) {
          copy.unit = 'percentunit';
          copy.display = getDisplayProcessor({ field: copy, theme });
        }
```
maybe?

**Comment by ryantxu** [comment]
behavior wise seems like progress!

image

Text placement is kind of a mess... but that is a whole can of worms:
image

**Review by ryantxu — APPROVED** [review]

File: public/app/plugins/panel/barchart/utils.ts (hunk 3 above)
Comment (leeoniya) [reply public/app/plugins/panel/barchart/utils.ts]: updated

**Pull request merged**